File: netIP.py
import os
import stat
import hashlib
import hmac
import secrets
from cryptography.hazmat.primitives.asymmetric import ed25519
from cryptography.hazmat.primitives import serialization
from cryptography.hazmat.primitives.ciphers.aead import AESGCM
from utils import load_env

load_env()
password = os.environ.get("MESH_PASSWORD", "").encode()


KEY_DIR = "./keys"
PRIVATE_FILE = os.path.join(KEY_DIR, "private.key")
PUBLIC_FILE  = os.path.join(KEY_DIR, "public.key")

ED25519_KEY_SIZE   = 32
HMAC_DERIVE_LENGTH = 32

# ----------------------------
# Sichere Dateierstellung mit korrekten Permissions von Anfang an
# FIX: Permissions werden VOR dem Schreiben gesetzt (kein TOCTOU-Fenster)
# ----------------------------
def _open_secure(path: str):
    """Öffnet/erstellt eine Datei mit 0o600 – atomisch, kein Race-Window."""
    fd = os.open(path, os.O_WRONLY | os.O_CREAT | os.O_TRUNC, 0o600)
    return os.fdopen(fd, "wb")

def _check_permissions(path: str) -> None:
    """Warnt, wenn die Datei für andere lesbar ist."""
    mode = os.stat(path).st_mode & 0o777
    if mode & 0o077:
        raise PermissionError(
            f"{path} ist zu offen ({oct(mode)}). Erwartet: 0o600."
        )

# ----------------------------
# Key laden oder erzeugen
# ----------------------------
def load_or_create_keys(password: bytes) -> tuple[ed25519.Ed25519PrivateKey, ed25519.Ed25519PublicKey]:
    os.makedirs(KEY_DIR, exist_ok=True)
    os.chmod(KEY_DIR, 0o700)

    if os.path.exists(PRIVATE_FILE) and os.path.exists(PUBLIC_FILE):
        return load_keys(password)  
    return create_keys(password)    

# ----------------------------
# Neue Keys erzeugen
# FIX: Private Key wird mit AES-256-GCM verschlüsselt gespeichert.
#      Das Passwort kommt hier als Parameter – nie hardcoden!
# ----------------------------
def create_keys(password: bytes) -> tuple[ed25519.Ed25519PrivateKey, ed25519.Ed25519PublicKey]:
    private_key = ed25519.Ed25519PrivateKey.generate()
    public_key  = private_key.public_key()

    raw_priv = private_key.private_bytes(
        encoding=serialization.Encoding.Raw,
        format=serialization.PrivateFormat.Raw,
        encryption_algorithm=serialization.NoEncryption(),  # Verschlüsselung machen wir selbst
    )
    raw_pub = public_key.public_bytes(
        encoding=serialization.Encoding.Raw,
        format=serialization.PublicFormat.Raw,
    )

    # FIX: Private Key mit AES-256-GCM + zufälligem Salt & Nonce verschlüsseln
    encrypted_priv = _encrypt_key(raw_priv, password)

    with _open_secure(PRIVATE_FILE) as f:
        f.write(encrypted_priv)
    with _open_secure(PUBLIC_FILE) as f:
        f.write(raw_pub)

    return private_key, public_key

# ----------------------------
# Keys laden
# FIX: Längenvalidierung der gelesenen Bytes
# ----------------------------
def load_keys(password: bytes) -> tuple[ed25519.Ed25519PrivateKey, ed25519.Ed25519PublicKey]:
    _check_permissions(PRIVATE_FILE)
    _check_permissions(PUBLIC_FILE)

    with open(PRIVATE_FILE, "rb") as f:
        encrypted_priv = f.read()
    with open(PUBLIC_FILE, "rb") as f:
        pub_bytes = f.read()

    # FIX: Größen prüfen bevor wir Bytes an Krypto-Funktionen übergeben
    if len(pub_bytes) != ED25519_KEY_SIZE:
        raise ValueError(f"Public Key hat ungültige Länge: {len(pub_bytes)}")

    raw_priv = _decrypt_key(encrypted_priv, password)
    if len(raw_priv) != ED25519_KEY_SIZE:
        raise ValueError(f"Private Key hat ungültige Länge nach Entschlüsselung: {len(raw_priv)}")

    priv = ed25519.Ed25519PrivateKey.from_private_bytes(raw_priv)
    pub  = ed25519.Ed25519PublicKey.from_public_bytes(pub_bytes)
    return priv, pub


def load_public_key() -> tuple[bytes, ed25519.Ed25519PublicKey]:
    """Lädt nur den Public Key – kein Passwort nötig."""
    _check_permissions(PUBLIC_FILE)
    with open(PUBLIC_FILE, "rb") as f:
        pub_bytes = f.read()
    if len(pub_bytes) != ED25519_KEY_SIZE:
        raise ValueError(f"Public Key hat ungültige Länge: {len(pub_bytes)}")
    return pub_bytes, ed25519.Ed25519PublicKey.from_public_bytes(pub_bytes)

# ----------------------------
# AES-256-GCM Hilfsfunktionen für Key-Verschlüsselung
# Format: salt(16) || nonce(12) || ciphertext+tag
# ----------------------------
def _derive_aes_key(password: bytes, salt: bytes) -> bytes:
    """PBKDF2-HMAC-SHA256: langsame KDF schützt gegen Brute-Force."""
    return hashlib.pbkdf2_hmac("sha256", password, salt, iterations=600_000, dklen=32)

def _encrypt_key(raw_key: bytes, password: bytes) -> bytes:
    salt  = secrets.token_bytes(16)
    nonce = secrets.token_bytes(12)
    aes_key = _derive_aes_key(password, salt)
    ct = AESGCM(aes_key).encrypt(nonce, raw_key, None)
    return salt + nonce + ct

def _decrypt_key(blob: bytes, password: bytes) -> bytes:
    if len(blob) < 16 + 12 + ED25519_KEY_SIZE + 16:  # salt+nonce+key+GCM-tag
        raise ValueError("Verschlüsselter Key-Blob ist zu kurz.")
    salt, nonce, ct = blob[:16], blob[16:28], blob[28:]
    aes_key = _derive_aes_key(password, salt)
    return AESGCM(aes_key).decrypt(nonce, ct, None)  # wirft InvalidTag bei Manipulation

# ----------------------------
# Adressableitung
# FIX 1: secrets statt random (random war kryptografisch unsicher)
# FIX 2: Längentrennzeichen verhindert Kollisionen (index="A", nonce="BC" ≠ index="AB", nonce="C")
# FIX 3: Kein Hostname mehr als Schlüssel (war vorhersehbar & niedrige Entropie)
# ----------------------------
def base_address(pub_bytes: bytes) -> str:
    if len(pub_bytes) != ED25519_KEY_SIZE:
        raise ValueError("Ungültige Public-Key-Länge.")
    return hashlib.sha256(pub_bytes).hexdigest()[:HMAC_DERIVE_LENGTH]

def derive_address(master_secret: bytes, index: bytes, nonce: bytes) -> str:
    """
    Leitet eine Adresse sicher per HMAC-SHA256 ab.
    master_secret: hochentropisches Secret (z.B. os.urandom(32)), KEIN Hostname
    Längenpräfixe verhindern Kollisionen zwischen (index, nonce)-Paaren.
    """
    # FIX: Längentrennzeichen: len(index) als 2-Byte-Präfix
    msg = len(index).to_bytes(2, "big") + index + nonce
    return hmac.new(master_secret, msg, hashlib.sha256).hexdigest()[:HMAC_DERIVE_LENGTH]

def gen_id(pub_bytes: bytes, length: int = 32) -> str:
    """
    Erzeugt eine zufällige, vom Public Key abgeleitete Adresse
    und kürzt sie auf die gewünschte Länge.

    Kein Privileg- oder Dateizugriff nötig.
    """
    if length <= 0:
        raise ValueError("length muss > 0 sein")
    if len(pub_bytes) != ED25519_KEY_SIZE:
        raise ValueError("Ungültige Public-Key-Länge")

    nonce = secrets.token_bytes(16)  # sorgt für Zufälligkeit
    digest = hashlib.sha256(pub_bytes + nonce).hexdigest()

    return digest[:length]


if __name__ == "__main__":
    
    master_secret = secrets.token_bytes(32)   
    nonce = secrets.token_bytes(16)           


    
    priv, pub = load_or_create_keys(password)
    pub_bytes = pub.public_bytes(serialization.Encoding.Raw, serialization.PublicFormat.Raw)

    print("Standard:", base_address(pub_bytes))
    # for i in range(5):
    #     addr = derive_address(master_secret, i.to_bytes(4, "big"), nonce)
    #     print(f"[{i}]", addr)

File: zig/netIP.zig
const std = @import("std");
const Io = std.Io;
const crypto = std.crypto;

const KEY_DIR = "./keys";
const PRIVATE_FILE = "keys/private.key";
const PUBLIC_FILE = "keys/public.key";
const PBKDF2_ITERATIONS = 600_000;

const KeyPair = struct {
    public: [32]u8,
    secret: [64]u8,
};

const KeyError = error{
    DecryptionFailed,
};

pub fn loadOrCreateKeys(io: std.Io, password: []const u8) !KeyPair {
    const cwd = Io.Dir.cwd();
    cwd.createDirPath(io, KEY_DIR) catch |err| switch (err) {
        error.PathAlreadyExists => {},
        else => return err,
    };
    const priv_exists = if (cwd.openFile(io, PRIVATE_FILE, .{})) |f| blk: {
        f.close(io);
        break :blk true;
    } else |_| false;
    const pub_exists = if (cwd.openFile(io, PUBLIC_FILE, .{})) |f| blk: {
        f.close(io);
        break :blk true;
    } else |_| false;
    if (priv_exists and pub_exists) {
        return loadKeys(io, password);
    } else {
        return createKeys(io, password);
    }
}

fn createKeys(io: std.Io, password: []const u8) !KeyPair {
    std.debug.print("[*] Erzeuge neue Schlüssel...\n", .{});

    const kp = crypto.sign.Ed25519.KeyPair.generate(io);
    const pub_bytes = kp.public_key.toBytes();
    const sec_bytes = kp.secret_key.toBytes();

    const rng_src: std.Random.IoSource = .{ .io = io };
    const rand = rng_src.interface();

    var salt: [16]u8 = undefined;
    var nonce: [crypto.aead.aes_gcm.Aes256Gcm.nonce_length]u8 = undefined;
    rand.bytes(&salt);
    rand.bytes(&nonce);

    var aes_key: [32]u8 = undefined;
    try crypto.pwhash.pbkdf2(&aes_key, password, &salt, PBKDF2_ITERATIONS, crypto.auth.hmac.sha2.HmacSha256);

    var encrypted_priv: [64]u8 = undefined;
    var tag: [crypto.aead.aes_gcm.Aes256Gcm.tag_length]u8 = undefined;
    crypto.aead.aes_gcm.Aes256Gcm.encrypt(&encrypted_priv, &tag, &sec_bytes, "", nonce, aes_key);

    const cwd = Io.Dir.cwd();
    {
        const f = try cwd.createFile(io, PRIVATE_FILE, .{});
        defer f.close(io);
        const rc = std.os.linux.syscall2(.fchmod, @intCast(f.handle), 0o600);
        if (rc != 0) return error.ChmodFailed;
        var buf: [256]u8 = undefined;
        var w = f.writer(io, &buf);
        try w.interface.writeAll(&salt);
        try w.interface.writeAll(&nonce);
        try w.interface.writeAll(&tag);
        try w.interface.writeAll(&encrypted_priv);
        try w.flush();
    }
    {
        // FIX: createFile statt openFile — public key schreiben nicht lesen
        const f = try cwd.createFile(io, PUBLIC_FILE, .{});
        defer f.close(io);
        const rc = std.os.linux.syscall2(.fchmod, @intCast(f.handle), 0o644);
        if (rc != 0) return error.ChmodFailed;
        var buf: [64]u8 = undefined;
        var w = f.writer(io, &buf);
        try w.interface.writeAll(&pub_bytes);
        try w.flush();
    }

    std.debug.print("[+] Keys gespeichert\n", .{});
    return KeyPair{ .public = pub_bytes, .secret = sec_bytes };
}

fn loadKeys(io: std.Io, password: []const u8) !KeyPair {
    std.debug.print("[*] Lade bestehende Schlüssel...\n", .{});

    var salt: [16]u8 = undefined;
    var nonce: [crypto.aead.aes_gcm.Aes256Gcm.nonce_length]u8 = undefined;
    var tag: [crypto.aead.aes_gcm.Aes256Gcm.tag_length]u8 = undefined;
    var encrypted_priv: [64]u8 = undefined;
    var pub_bytes: [32]u8 = undefined;

    const cwd = Io.Dir.cwd();
    {
        const f = try cwd.openFile(io, PRIVATE_FILE, .{});
        defer f.close(io);
        var raw: [108]u8 = undefined;
        _ = try f.readPositionalAll(io, &raw, 0);
        salt = raw[0..16].*;
        nonce = raw[16..28].*;
        tag = raw[28..44].*;
        encrypted_priv = raw[44..108].*;
    }
    {
        const f = try cwd.openFile(io, PUBLIC_FILE, .{});
        defer f.close(io);
        _ = try f.readPositionalAll(io, &pub_bytes, 0);
    }

    var aes_key: [32]u8 = undefined;
    try crypto.pwhash.pbkdf2(&aes_key, password, &salt, PBKDF2_ITERATIONS, crypto.auth.hmac.sha2.HmacSha256);

    var sec_bytes: [64]u8 = undefined;
    crypto.aead.aes_gcm.Aes256Gcm.decrypt(&sec_bytes, &encrypted_priv, tag, "", nonce, aes_key) catch {
        std.debug.print("[!] Falsches Passwort oder beschädigte Keys\n", .{});
        return KeyError.DecryptionFailed;
    };

    return KeyPair{ .public = pub_bytes, .secret = sec_bytes };
}

pub fn baseAddress(pub_bytes: [32]u8) [32]u8 {
    var out: [crypto.hash.sha2.Sha256.digest_length]u8 = undefined;
    crypto.hash.sha2.Sha256.hash(&pub_bytes, &out, .{});
    return out;
}

pub fn genId(io: std.Io, pub_bytes: [32]u8) [32]u8 {
    var nonce: [16]u8 = undefined;
    const rng_src: std.Random.IoSource = .{ .io = io };
    rng_src.interface().bytes(&nonce);

    var h = crypto.hash.sha2.Sha256.init(.{});
    h.update(&pub_bytes);
    h.update(&nonce);
    var out: [32]u8 = undefined;
    h.final(&out);
    return out;
}

pub fn main(init: std.process.Init) !void {
    const io = init.io;
    const password = "MeinSuperSicheresMeshPasswort!";
    const kp = try loadOrCreateKeys(io, password);
    const addr = baseAddress(kp.public);
    std.debug.print("Public Key   : {x}\n", .{kp.public});
    std.debug.print("Basis-Addr   : {x}\n", .{addr});
    const id = genId(io, kp.public);
    std.debug.print("Generierte ID: {x}\n", .{id});
}
